source/mir/random/discrete.d

```diff
@@ -0,0 +1,123 @@
+/**
+Discrete distributions.
+
+License: $(LINK2 http://boost.org/LICENSE_1_0.txt, Boost License 1.0).
+
+Authors: Ilya Yaroshenko, Sebastian Wilzbach
+*/
+
+module mir.random.discrete;
+
+import std.traits : isNumeric;
+
+/**
+Setup a _discrete distribution sampler.
+Given an array of cumulative density points `cdPoints`,
+a _discrete distribution is defined.
+The cumulative density points can be calculated from probabilities or counts
+using `std.algorithm.iteration.cumulativeFold`. Hence `cdPoints` is assumed to be sorted.
+
+Params:
+    cdPoints = cumulative density points
+
+Returns:
+    A $(LREF Discrete) sampler that can be called to sample from the distribution.
+*/
+Discrete!T discrete(T)(const(T)[] cdPoints)
+{
+    return Discrete!T(cdPoints);
+}
+
+///
+unittest
+{
+    // 10%, 20%, 20%, 40%, 10%
+    auto cdPoints = [0.1, 0.3, 0.5, 0.9, 1];
+    auto ds = discrete(cdPoints);
+
+    // sample from the discrete distribution
+    auto obs = new uint[cdPoints.length];
+    foreach (i; 0..1000)
+        obs[ds()]++;
+}
+
+/**
+_Discrete distribution sampler that draws random values from a _discrete
+distribution given an array of the respective cumulative density points.
+`cdPoints` is an array of the cummulative sum over the probabilities
+or counts of a _discrete distribution without a starting zero.
+
+Complexity: O(log n) where n is the number of `cdPoints`.
+*/
+struct Discrete(T)
+    if (isNumeric!T)
+{
+    import std.range : SortedRange;
+
+    private const(T)[] _cdPoints;
+    private SortedRange!(const(T)[], "a <= b") r;
```

Comment:
> @wilzbach fix this please. SortedRange should have something like release

`release` uses `move`, so that won't work? We can just remove `cdPoints` from the public interface?

https://github.com/dlang/phobos/blob/master/std/range/package.d#L8023